src/history/hs.c

```diff
@@ -1071,3 +1088,72 @@ __wt_find_hs_upd(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE **upd
 
     return (ret);
 }
+
+/*
+ * __hs_delete_key --
+ *     Delete an entire key's worth of data in the history store.
+ */
+static int
+__hs_delete_key(
+  WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id, const WT_ITEM *key)
+{
+    WT_CURSOR_BTREE *hs_cbt;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    WT_TIME_PAIR hs_start, hs_stop;
+    WT_UPDATE *upd;
+    size_t size;
+    uint32_t hs_btree_id;
+    int cmp;
+
+    hs_cbt = (WT_CURSOR_BTREE *)hs_cursor;
+    upd = NULL;
+
+#ifdef HAVE_DIAGNOSTIC
+    /*
+     * If we've decided we need to delete a key from the history store, we should have JUST inserted
+     * a zero timestamp update into the history store. Assuming this, we can just keep iterating
+     * until we hit the key boundary, inserting tombstones as we go.
+     */
+    WT_RET(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_start.timestamp,
+      &hs_start.txnid, &hs_stop.timestamp, &hs_stop.txnid));
+    WT_ASSERT(session, hs_btree_id == btree_id);
+    WT_RET(__wt_compare(session, NULL, &hs_key, key, &cmp));
+    WT_ASSERT(session, cmp == 0);
+    WT_ASSERT(session, hs_start.timestamp == 0);
+#endif
+    /* If there is nothing else in history store, we're done here. */
+    ret = hs_cursor->next(hs_cursor);
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
+        WT_RET(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_start.timestamp,
+          &hs_start.txnid, &hs_stop.timestamp, &hs_stop.txnid));
+        /*
+         * If the btree id or key isn't ours, that means that we've hit the end of the key range and
+         * that there is no more history store content for this key.
+         */
+        if (hs_btree_id != btree_id)
+            break;
+        WT_RET(__wt_compare(session, NULL, &hs_key, key, &cmp));
+        if (cmp != 0)
+            break;
+        /* Ignore non-timestamped writes. These won't cause visibility issues. */
+        if (hs_start.timestamp == WT_TS_NONE)
+            continue;
+        /*
+         * Append a globally visible tombstone to the update list. This will effectively make the
+         * value invisible and the key itself will eventually get removed during reconciliation.
+         */
+        WT_RET(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_NONE;
+        upd->start_ts = upd->durable_ts = WT_TS_NONE;
+        WT_WITH_BTREE(session, hs_cbt->btree,
+          ret = __wt_row_modify(hs_cbt, &hs_cursor->key, NULL, upd, WT_UPDATE_INVALID, true));
+        WT_ERR(ret);
+        upd = NULL;
+    }
+    if (ret == WT_NOTFOUND)
+        return (0);
+err:
+    __wt_free_update_list(session, &upd);
```

Comment:
There is only one `upd` can we directly free it using `__wt_free()`?

Reply:
197c295

Done. However, we use `__wt_free_update_list` to free a single update in a lot of places. The most obvious example is when we free an update instantiated from the history store.